pkg/operator/configobservation/audit/observe_audit_policy.go

```diff
@@ -0,0 +1,74 @@
+package audit
+
+import (
+	gojson "encoding/json"
+	"fmt"
+
+	yaml2 "sigs.k8s.io/yaml"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/configobservation"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+)
+
+const policyConfigMapName = "audit-policy"
+
+// ObserveAuditPolicy observes the openshift-config/audit-policy ConfigMap and merges it into the kube-apiserver config.
+func ObserveAuditPolicy(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	errs := []error{}
+	prevObservedConfig := map[string]interface{}{}
+
+	// copy non-empty .auditConfig.policyConfiguration from existingConfig to prevObservedConfig
+	policyPath := []string{"auditConfig", "policyConfiguration"}
+	existingPolicy, _, err := unstructured.NestedFieldNoCopy(existingConfig, policyPath...)
+	if err != nil {
+		return prevObservedConfig, append(errs, err)
+	}
+	if existingPolicy != nil {
+		if err := unstructured.SetNestedField(prevObservedConfig, existingPolicy, policyPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	observedConfig := map[string]interface{}{}
+
+	// add initial-sa-token-signing-certs configmap mount path to saTokenCertDirs if configmap exists
+	policyConfigMap, err := listers.ConfigmapLister.ConfigMaps(operatorclient.GlobalUserSpecifiedConfigNamespace).Get(policyConfigMapName)
+	switch {
+	case errors.IsNotFound(err):
+		// switch back to default policy by leaving observedConfig empty
+	case err != nil:
+		// we had an error, return what we had before and exit. this really shouldn't happen
+		return prevObservedConfig, append(errs, err)
+	case len(policyConfigMap.Data["policy.yaml"]) > 0:
+		// TODO: maybe verify config here
```

Comment:
You mean you'd like to verify before applying it to server to ensure it doesn't crash the server/rejected by the server?

Reply:
yes, return an early error. But that binds the operator livecycle to that of the server. Maybe not desired.

Reply:
I don't think we do that kind of validation elsewhere in the operator code, so I'd do it here until proven we need it.